File: apps/api/schema/student.ts
import { InferInsertModel, InferSelectModel, relations, sql } from 'drizzle-orm'
import * as sqlite from 'drizzle-orm/sqlite-core'
import { customType } from 'drizzle-orm/sqlite-core'
import { AppError } from '../errors/index'
import { baseSchema } from './base'
import { PositionDB, positions } from './positions'
import { Unit, units } from './units'

const PoliticalOrgEnum = customType<{ data: string; driverData: string }>({
	dataType() {
		return 'text'
	},
	toDriver(val: string) {
		if (!['hcyu', 'cpv'].includes(val)) {
			throw AppError.invalidArgument(
				'politicalOrg can be only hcyu or cpv'
			)
		}
		return val
	}
})

export const students = sqlite.sqliteTable('students', {
	...baseSchema,
	fullName: sqlite.text().default(''),
	birthPlace: sqlite.text().default(''),
	address: sqlite.text().default(''),
	dob: sqlite.text().default(''),
	rank: sqlite.text().default(''),
	previousUnit: sqlite.text().default(''),
	previousPosition: sqlite.text().default(''),
	// Derived, read-only mirror of positionId's `code` - set automatically
	// by students/controller.ts#resolvePositionText whenever positionId is
	// given. Kept as a real column (rather than dropped) because
	// roster-utils.ts's export sorting and validateUniqueLeaderPositions
	// both key off this text, not the FK.
	position: sqlite.text().default('Chiến sĩ'),
	positionId: sqlite.integer().references(() => positions.id),
	ethnic: sqlite.text().default(''),
	religion: sqlite.text().default('Không'),
	enlistmentPeriod: sqlite.text().default(''),
	politicalOrg: PoliticalOrgEnum('politicalOrg')
		.$type<'hcyu' | 'cpv'>()
		.notNull(),
	politicalOrgOfficialDate: sqlite.text().default(''),
	cpvId: sqlite.text(),
	educationLevel: sqlite.text().default(''),
	schoolName: sqlite.text().default(''),
	major: sqlite.text().default(''),
	isGraduated: sqlite.int({ mode: 'boolean' }).default(false),
	talent: sqlite.text().default('Không'),
	shortcoming: sqlite.text().default('Không'),
	policyBeneficiaryGroup: sqlite.text().default('Không'),
	fatherName: sqlite.text().default(''),
	fatherDob: sqlite.text().default(''),
	fatherPhoneNumber: sqlite.text().default(''),
	fatherJob: sqlite.text().default(''),
	// fatherJobAddress: sqlite.text().default(''),
	motherName: sqlite.text().default(''),
	motherDob: sqlite.text().default(''),
	motherPhoneNumber: sqlite.text().default(''),
	motherJob: sqlite.text().default(''),
	// motherJobAddress: sqlite.text().default(''),
	isMarried: sqlite.int({ mode: 'boolean' }).default(false),
	spouseName: sqlite.text().default(''),
	spouseDob: sqlite.text().default(''),
	spouseJob: sqlite.text().default(''),
	spousePhoneNumber: sqlite.text().default(''),
	childrenInfos: sqlite.text({ mode: 'json' }).default(sql`'[]'`),
	familySize: sqlite.int(),
	familyBackground: sqlite.text().default('Không'),
	familyBirthOrder: sqlite.text().default(''),
	achievement: sqlite.text().default('Không'),
	disciplinaryHistory: sqlite.text().default('Không'),
	phone: sqlite.text().default(''),
	unitId: sqlite.integer().references(() => units.id),
	cpvOfficialAt: sqlite.text(),
	avatar: sqlite.text(),
	siblings: sqlite.text({ mode: 'json' }).default(sql`'[]'`),
	contactPerson: sqlite.text({ mode: 'json' }).default(sql`'{}'`),
	studentId: sqlite.text(),
	relatedDocumentations: sqlite.text(),
	status: sqlite
		.text()
		.$type<'pending' | 'confirmed'>()
		.default('pending')
		.notNull()
})

export const studentsRelations = relations(students, ({ one }) => ({
	unit: one(units, {
		fields: [students.unitId],
		references: [units.id]
	}),
	positionRef: one(positions, {
		fields: [students.positionId],
		references: [positions.id]
	})
}))

export type StudentDB = InferSelectModel<typeof students>

export type Student = Omit<StudentDB, 'unitId' | 'positionId'> & {
	unit: Unit | null
	positionRef: PositionDB | null
}

export type StudentParam = InferInsertModel<typeof students>

export type Month =
	| '01'
	| '02'
	| '03'
	| '04'
	| '05'
	| '06'
	| '07'
	| '08'
	| '09'
	| '10'
	| '11'
	| '12'

export type Quarter = 'Q1' | 'Q2' | 'Q3' | 'Q4'

export type StudentQuery = {
	birthdayInMonth?: Month
	birthdayInQuarter?: Quarter
	birthdayInWeek?: boolean
	unitIds?: Array<number>
	hasReligion?: boolean
	ids?: Array<number>
	isEthnicMinority?: boolean
	isMarried?: boolean
	politicalOrg?: 'hcyu' | 'cpv'
	isCpvOfficialThisWeek?: boolean
	cpvOfficialInMonth?: Month
	cpvOfficialInQuarter?: Quarter
	withAdversity?: boolean
}

export type UpdateStudentMap = {
	id: number
	updatePayload: { [k: string]: any }
}[]

export interface StudentCronJobEvent {
	getQueryParams(): StudentQuery
}

export class BirthdayThisWeek implements StudentCronJobEvent {
	constructor() {}

	getQueryParams(): StudentQuery {
		return { birthdayInWeek: true }
	}
}

export class BirthdayThisMonth implements StudentCronJobEvent {
	constructor(private month: Month) {}

	getQueryParams(): StudentQuery {
		return { birthdayInMonth: this.month }
	}
}

export class BirthdayThisQuarter implements StudentCronJobEvent {
	constructor(private quarter: Quarter) {}

	getQueryParams(): StudentQuery {
		return { birthdayInQuarter: this.quarter }
	}
}

export class CpvOfficialThisWeek implements StudentCronJobEvent {
	constructor() {}

	getQueryParams(): StudentQuery {
		return { isCpvOfficialThisWeek: true }
	}
}

export class CpvOfficialThisMonth implements StudentCronJobEvent {
	constructor(private month: Month) {}

	getQueryParams(): StudentQuery {
		return { cpvOfficialInMonth: this.month }
	}
}

export class CpvOfficialThisQuarter implements StudentCronJobEvent {
	constructor(private quarter: Quarter) {}

	getQueryParams(): StudentQuery {
		return { cpvOfficialInQuarter: this.quarter }
	}
}

export type StudentCronEvent =
	| 'birthdayThisWeek'
	| 'birthdayThisMonth'
	| 'birthdayThisQuarter'
	| 'cpvOfficialThisWeek'
	| 'cpvOfficialThisMonth'
	| 'cpvOfficialThisQuarter'

export type PoliticsQualityRow = {
	category: string
	unitId: number
	value: string | number
	count: number
}

export type ExcelTemplateData = {
	city: string
	commanderName: string
	commanderPosition: string
	commanderRank: string
	day: string
	month: string
	rows: Record<string, string>[]
	underUnitName: string
	unitName: string
	year: number
}

export const templateTypes = [
	'CpvTempl',
	'HcyuTempl',
	'StudentInfoTempl',
	'StudentWithAdversityTempl',
	'StudentEnrollmentFormTempl'
] as const

export type TemplateType = (typeof templateTypes)[number]

File: apps/api-go/internal/entities/student.go
// Package entities holds the persistence-layer structs shared between the
// students repository and its API layer.
package entities

// PoliticalOrg mirrors schema/student.ts's PoliticalOrgEnum custom type.
type PoliticalOrg string

const (
	PoliticalOrgHCYU PoliticalOrg = "hcyu"
	PoliticalOrgCPV  PoliticalOrg = "cpv"
)

// Status mirrors schema/student.ts's status enum.
type Status string

const (
	StatusPending   Status = "pending"
	StatusConfirmed Status = "confirmed"
)

// Student is the students table row. Field order and `db` tags match the
// SQLite columns created by apps/api's Drizzle migrations, so this struct
// can read/write the exact same table.
type Student struct {
	ID        int64  `json:"id"                  db:"pk"`
	CreatedAt string `json:"createdAt,omitempty" db:"createdAt"`
	UpdatedAt string `json:"updatedAt,omitempty" db:"updatedAt"`

	FullName                 string       `json:"fullName"                 db:"fullName"`
	BirthPlace               string       `json:"birthPlace"               db:"birthPlace"`
	Address                  string       `json:"address"                  db:"address"`
	Dob                      string       `json:"dob"                      db:"dob"`
	Rank                     string       `json:"rank"                     db:"rank"`
	PreviousUnit             string       `json:"previousUnit"             db:"previousUnit"`
	PreviousPosition         string       `json:"previousPosition"         db:"previousPosition"`
	Position                 string       `json:"position"                 db:"position"`
	Ethnic                   string       `json:"ethnic"                   db:"ethnic"`
	Religion                 string       `json:"religion"                 db:"religion"`
	EnlistmentPeriod         string       `json:"enlistmentPeriod"         db:"enlistmentPeriod"`
	PoliticalOrg             PoliticalOrg `json:"politicalOrg"             db:"politicalOrg"`
	PoliticalOrgOfficialDate string       `json:"politicalOrgOfficialDate" db:"politicalOrgOfficialDate"`
	CpvID                    *string      `json:"cpvId,omitempty"          db:"cpvId"`
	EducationLevel           string       `json:"educationLevel"           db:"educationLevel"`
	SchoolName               string       `json:"schoolName"               db:"schoolName"`
	Major                    string       `json:"major"                    db:"major"`
	IsGraduated              bool         `json:"isGraduated"              db:"isGraduated"`
	Talent                   string       `json:"talent"                   db:"talent"`
	Shortcoming              string       `json:"shortcoming"              db:"shortcoming"`
	PolicyBeneficiaryGroup   string       `json:"policyBeneficiaryGroup"   db:"policyBeneficiaryGroup"`
	FatherName               string       `json:"fatherName"               db:"fatherName"`
	FatherDob                string       `json:"fatherDob"                db:"fatherDob"`
	FatherPhoneNumber        string       `json:"fatherPhoneNumber"        db:"fatherPhoneNumber"`
	FatherJob                string       `json:"fatherJob"                db:"fatherJob"`
	MotherName               string       `json:"motherName"               db:"motherName"`
	MotherDob                string       `json:"motherDob"                db:"motherDob"`
	MotherPhoneNumber        string       `json:"motherPhoneNumber"        db:"motherPhoneNumber"`
	MotherJob                string       `json:"motherJob"                db:"motherJob"`
	IsMarried                bool         `json:"isMarried"                db:"isMarried"`
	SpouseName               string       `json:"spouseName"               db:"spouseName"`
	SpouseDob                string       `json:"spouseDob"                db:"spouseDob"`
	SpouseJob                string       `json:"spouseJob"                db:"spouseJob"`
	SpousePhoneNumber        string       `json:"spousePhoneNumber"        db:"spousePhoneNumber"`
	// ChildrenInfos/Siblings/ContactPerson are stored as raw JSON text
	// (matching the "text" column affinity apps/api's Drizzle schema
	// declares), not driver-level BLOBs, so apps/api can keep reading them
	// as JSON strings.
	ChildrenInfos         JSONText `json:"childrenInfos"            db:"childrenInfos"`
	FamilySize            *int64   `json:"familySize,omitempty"     db:"familySize"`
	FamilyBackground      string   `json:"familyBackground"         db:"familyBackground"`
	FamilyBirthOrder      string   `json:"familyBirthOrder"         db:"familyBirthOrder"`
	Achievement           string   `json:"achievement"              db:"achievement"`
	DisciplinaryHistory   string   `json:"disciplinaryHistory"      db:"disciplinaryHistory"`
	Phone                 string   `json:"phone"                    db:"phone"`
	ClassID               *int64   `json:"classId,omitempty"        db:"classId"`
	UnitID                *int64   `json:"unitId,omitempty"         db:"unitId"`
	CpvOfficialAt         *string  `json:"cpvOfficialAt,omitempty"  db:"cpvOfficialAt"`
	Avatar                *string  `json:"avatar,omitempty"         db:"avatar"`
	Siblings              JSONText `json:"siblings"                 db:"siblings"`
	ContactPerson         JSONText `json:"contactPerson"            db:"contactPerson"`
	StudentID             *string  `json:"studentId,omitempty"      db:"studentId"`
	RelatedDocumentations *string  `json:"relatedDocumentations,omitempty" db:"relatedDocumentations"`
	Status                Status   `json:"status"                   db:"status"`
}

// TableName implements dbx.TableModel — the struct name would otherwise be
// mapped to "student", not the actual "students" table.
func (Student) TableName() string {
	return "students"
}
